File: python/tree-building/tree_building.py
"""
Rebuild a tree structure from a list of parent/child ids
"""

class Record:
    """
    Record class holds data that would have been retrieved from the data store with
    record and its parent relationship.
    """
    def __init__(self, record_id, parent_id):
        """
        Initialize the Record object with the given record and parent IDs
        :param record_id: id to set this record to.
        :param parent_id: id of this records parent (0 if root)
        """
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    """
    Tree Node class
    """
    def __init__(self, node_id):
        """
        Create a new tree node giving it the passed in id.
        Children will be initialized to an empty list.
        :param node_id: Value to set this node's node_id to
        """
        self.node_id = node_id
        self.children = []

def BuildTree(records):
    """
    Reconstruct a tree of nodes from the given records.
    :param records: A list of record objects to reconstruct the tree with.
    :returns: root of the reconstructed tree or None if there are no nodes.
    :raises: Value error if there are improper parent/child relationships, missing
    records/records not in sequential order or similar data that doesn't meet expectations.
    """
    record_ids = [record.record_id for record in records]
    if records:
        min_id = min(record_ids)
        max_id = max(record_ids)
        if max_id != len(records) - 1:
            raise ValueError('Record id is invalid or out of order.')
        if min_id != 0:
            raise ValueError('Record id is invalid or out of order.')
    trees = []
    for record in records:    
        if record.record_id == 0 and record.parent_id != 0:
            raise ValueError('Node parent_id should be smaller than its record_id.') 
        if record.record_id < record.parent_id:
            raise ValueError('Node parent_id should be smaller than its record_id.')
        if record.record_id == record.parent_id and record.record_id != 0:
            raise ValueError('Only root should have equal record and parent id.') 
        trees.append(Node(record.record_id))
    
    for tree in trees:
        child_ids = [record.record_id for record in records if record.parent_id == tree.node_id and record.record_id != tree.node_id]
        children = [node for node in trees if node.node_id in child_ids]
        children.sort(key=lambda x: x.node_id)
        tree.children.extend(children)

    roots = [tree for tree in trees if tree.node_id == 0] #trees[0]
    if len(roots) == 0:
        return None
    return roots[0]

File: go/tree-building/tree_building.go
// Tree Building exercise. I thought we were supposed to be refactoring.
package treebuilding

import (
	"errors"
	"slices"
)

// I thought we were supposed to be refactoring.

type Record struct {
	ID     int
	Parent int
	// feel free to add fields as you see fit
}

type Node struct {
	ID       int
	Children []*Node
	// feel free to add fields as you see fit
}

// Build a tree of nodes from a slice of records.
// @param records: slice of record.
// @returns: Root node of the reconstituted tree or an error.
// @raises: Error for node IDs not in order or duplicated, bad parent ID, multiple root nodes
// or lookup for node, node's parent, or root node not found.
func Build(records []Record) (*Node, error) {
	nodes := make(map[int]*Node, 0)
	// Make a node for each record.
	for _, rec := range records {
		if rec.Parent > rec.ID {
			return nil, errors.New("Parent ID must be less than or equal to ID")
		}
		if rec.ID >= len(records) {
			return nil, errors.New("ID is too large")
		}
		nodes[rec.ID] = &Node{ID: rec.ID, Children: []*Node{}}
	}

	if len(nodes) < len(records) {
		return nil, errors.New("Duplicate ID found.")
	}

	// Build the tree
	var root *Node = nil
	var ok bool
	for _, rec := range records {
		if rec.ID == rec.Parent {
			if root != nil {
				return nil, errors.New("Multiple root nodes found.")
			}
			root, ok = nodes[rec.ID]
			if !ok {
				return nil, errors.New("Expected node not found - root.")
			}
		} else {
			node, ok := nodes[rec.ID]
			if !ok {
				return nil, errors.New("Expected node not found - child.")
			}
			parent, ok := nodes[rec.Parent]
			if !ok {
				return nil, errors.New("Expected node not found - parent.")
			}
			parent.Children = append(parent.Children, node)
			slices.SortFunc(parent.Children, func(a, b *Node) int { return a.ID - b.ID })
		}
	}
	if root == nil {
		return nil, errors.New("No root node found.")
	}
	return root, nil
}
